text/0000-generalized-arity-tuples.md

```diff
@@ -0,0 +1,277 @@
+
+## Why are we already done here?
+
+Consider the type `(Option<i32>, (Option<bool>, Option<&str>))` from the requirements above. This type is just syntactic sugar for the type `Tuple<Option<i32>, Tuple<Tuple<Option<bool>, Tuple<Option<&str>, ()>>, ()>>`.
+
+Now, apply the provided trait implementations step by step on the desugared type. The resulting associated type `Join::Joined` evaluates to `Tuple<i32, Tuple<Tuple<bool, Tuple<&str, ()>>, ()>>`. But this is just the desugared version of `(i32, (bool, &str))`.
+
+This illustrates how the mechanims works on the type level.
+
+## More examples / Advanced type mappings
+
+Generalized implementations on `Tuple<HEAD, TAIL>` are not restricted to mappings between tuples of the same shape. The following example demonstrates how a `last` function can be realized that works with any tuple:
+
+```rust
+trait Last {
+    type Last;
+
+    fn last(self) -> Self::Last;
+}
+
+impl<ELEM> Last for (ELEM,) {
+    type Last = ELEM;
+
+    fn last(self) -> Self::Last {
+        self.elem
+    }
+}
+
+impl<ELEM, TAIL: Last> Last for Tuple<ELEM, TAIL> {
+    type Last = TAIL::Last;
+
+    fn last(self) -> Self::Last {
+        self.tail.last()
+    }
+}
+```
+
+Correcty, the compiler rejects empty tuples while tuples of other sizes are accepted:
+
+```rust
+().last();       // does not compile: no method named `last` found for type `()` in the current scope
+(1,).last();     // returns 1
+(1, "two").last(); // returns "two"
+etc.
+```
+
+The last example demonstrates how every second element of a tuple can be removed:
+
+```rust
+trait Halve {
+    type Output;
+
+    fn halve(self) -> Self::Output;
+}
+
+impl Halve for () {
+    type Output = ();
+
+    fn halve(self) {}
+}
+
+impl<ELEM1, ELEM2, TAIL: Halve> Halve for Tuple<ELEM1, Tuple<ELEM2, TAIL>> {
+    type Output = Tuple<ELEM1, TAIL>;
+
+    fn halve(self) -> Self::Output {
+        Tuple::new(self.elem, self.tail.tail)
+    }
+}
+```
+
+Results:
+
+```rust
+().halve();                   // returns ()
+(1,).halve()                  // does not compile: no method named `halve` found for type `(i32,)` in the current scope
+(1, "two").halve();           // returns (1,)
+(1, "two", 3.0, '4').halve(); // returns (1, 3.0)
+etc.
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The idea of this RFC to leave the tuple notation as is but to treat tuple type expressions as aliases according to the following pattern:
+
+```rust
+type () = (); // Not really an alias. Written down for completeness.
+type (A,) = Tuple<A, ()>;
+type (A, B) = Tuple<A, Tuple<B, ()>>;
+type (A, B, C) = Tuple<A, Tuple<B, Tuple<C, ()>>>;
+type (A, (B, C)) = Tuple<A, Tuple<Tuple<B, Tuple<C, ()>>, ()>>
+etc.
+```
+
+where `Tuple` is a new struct located in `std::ops` with the following definition:
+
+```rust
+struct Tuple<ELEM, TAIL> {
+    pub elem: ELEM,
+    pub tail: TAIL,
+}
+```
```

Comment:
Is this compatible with [the layout specified in the unsafe code guidelines](https://rust-lang.github.io/unsafe-code-guidelines/layout/structs-and-tuples.html#tuple-types)? 

> an anonymous tuple type (T1..Tn) of arity N is laid out "as if" there were a corresponding tuple struct declared in libcore:
>
> ```rust
> #[repr(Rust)]
> struct TupleN<P1..Pn:?Sized>(P1..Pn);
> ```

Note that this specifies that the layout of a tuple does not match that of the struct you propose, but of a tuple-struct. 

This allows the compiler to perform some optimizations, like field re-ordering, but this RFC does not mention anything about this particular trade-off. I'd recommend scanning the [UCGs repo](https://github.com/rust-lang/unsafe-code-guidelines/issues?utf8=%E2%9C%93&q=tuple+layout) for trade-offs and extending the RFC with how would this change affect those.